File: flow/src/main/java/flow/FlowModelManager.java
package flow;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

class FlowModelManager {
  private final List<FlowModelScope> scopes;

  FlowModelManager(@NonNull List<FlowModelScope> scopes) {
    this.scopes = scopes;
  }

  Object getModel(@NonNull Class scopeClass, @NonNull String tag) {
    FlowModelScope scope = getScope(scopeClass);
    if (scope == null) {
      return null;
    }

    return scope.getModel(tag);
  }

  void setUp(@NonNull FlowModelUser key) {
    for (Class scopeClass: key.getRelations().getScopes()) {
      FlowModelScope scope = getScope(scopeClass);
      if (scope == null) {
        continue;
      }

      scope.setUp(key);
    }
  }

  void tearDown(@NonNull FlowModelUser key) {
    for (Class scopeClass: key.getRelations().getScopes()) {
      FlowModelScope scope = getScope(scopeClass);
      if (scope == null) {
        return;
      }

      scope.tearDown(key);
	}
  }

  @Nullable
  private FlowModelScope getScope(@NonNull Class scopeClass) {
    for (FlowModelScope scope : scopes) {
      if (scope.getClass().isAssignableFrom(scopeClass)) {
        return scope;
      }
    }

    return null;
  }

}

File: flow/src/main/java/flow/FlowModelManager.kt
package flow

@Mockable
class FlowModelManager(
  private val scopes: List<FlowModelScope>
) {

  fun getModel(scopeClass: Class<*>, tag: String): Any? =
    findScope(scopeClass)?.getModel(tag)

  fun setUp(key: FlowModelUser) {
    for (scopeClass in key.relations.scopes) {
      findScope(scopeClass)?.setUp(key)
    }
  }

  fun tearDown(key: FlowModelUser) {
    for (scopeClass in key.relations.scopes) {
      findScope(scopeClass)?.tearDown(key)
    }
  }

  private fun findScope(scopeClass: Class<*>): FlowModelScope? =
    scopes.firstOrNull { it.javaClass.isAssignableFrom(scopeClass) }

}
